### DISCUSSION TITLE
How tol solve this DRY violation

### DISCUSSION THREAD

**Discussion by ivan-kleshnin** [colinhacks/zod#822] [Q&A]
It's probably more a TS question but it's related to Zod and UI and looks like a common case that should be interesting to someone else.

Currently, I have to declare two objects with the same string constants because I don't know how to derive one from another:

```ts
const OPPORTUNITY_OPTIONS = {"ClientList": "List", "Event": "Event"} as const // need this for Select options
const OPPORTUNITY_TYPES = ["ClientList", "Event"] as const // need this for Zod enum
type OpportunityType = typeof OPPORTUNITY_TYPES[number]
// OR type OpportunityType = keyof (typeof OPPORTUNITY_OPTIONS)

...
const schema = Z.object({
  opportunityType: Z.enum(OPPORTUNITY_TYPES), // can't derive a proper value from OPPORTUNITY_OPTIONS
```

Now imagine there are 20 options. It becomes a problem.
Is there any way to derive `OPPORTUNITY_TYPES` from `OPPORTUNITY_OPTIONS`?
I don't see any viable option to go from `Object.keys / union` to the tuple of strings.

Automatic Union -> Tuple conversion, while possible, is extremely CPU and RAM intensive so it's not a scalable solution:
https://github.com/microsoft/TypeScript/issues/13298

The only approach I see right now it to update or wrap `enum` to accept both tuples & objects and use object keys if the object was passed. But if someone case suggest something better – I'm eager to know.

Sorry if it was asked previously.

**Comment by JacobWeisenburger** [comment]
Hope this helps.

```ts
const OPPORTUNITY_OPTIONS = { ClientList: 'List', Event: 'Event' } as const

type OpportunityType = keyof typeof OPPORTUNITY_OPTIONS
//-> OpportunityType = "Event" | "ClientList"

type OpportunityTypes = TuplifyUnion<OpportunityType>
//-> OpportunityTypes = ["Event", "ClientList"]

const OPPORTUNITY_TYPES = Object.keys( OPPORTUNITY_OPTIONS ) as OpportunityTypes

const schema = z.object( {
    opportunityType: z.enum( OPPORTUNITY_TYPES ),
} )
```

[stackoverflow: How to transform union type to tuple type](https://stackoverflow.com/a/55128956/5321863)
```ts
type UnionToIntersection<U> =
  (U extends any ? (k: U) => void : never) extends ((k: infer I) => void) ? I : never

type LastOf<T> =
  UnionToIntersection<T extends any ? () => T : never> extends () => (infer R) ? R : never

type Push<T extends any[], V> = [...T, V];

type TuplifyUnion<T, L = LastOf<T>, N = [T] extends [never] ? true : false> =
  true extends N ? [] : Push<TuplifyUnion<Exclude<T, L>>, L>
```
___
I was able to get up to 48 properties before TS gave me the error:
`Type instantiation is excessively deep and possibly infinite.`

```ts
type StressTest = TuplifyUnion<keyof {
    '1': 'string value 1',
    '2': 'string value 2',
    '3': 'string value 3',
    ...
    '46': 'string value 46',
    '47': 'string value 47',
    '48': 'string value 48',
}>
//-> StressTest = ["1", "2", "3", ..., "46", "47", "48"]
```

```ts
type StressTest = TuplifyUnion<keyof {
    //            ^^^^^^^^^^^^^^^^^^^^
    // Type instantiation is excessively deep and possibly infinite
    '1': 'string value 1',
    '2': 'string value 2',
    '3': 'string value 3',
    ...
    '47': 'string value 47',
    '48': 'string value 48',
    '49': 'string value 49',
}>
```

**Comment by ivan-kleshnin** [comment]
@JacobWeisenburger thank you for the response! I think I saw this dark magic from Jcalz 😅  but I didn't check how many items it actually supports. And "almost 50" is a solid length 👍  For now I think I'd probably still prefer a custom `Z.enum` version, as mentioned about, but what you describe is also viable.

**Reply by JacobWeisenburger** [reply]
Sounds good. But can I ask why you prefer a custom `Z.enum` version? And what is not preferable about the way I described? I'm not trying to argue or anything, just wanting to understand.

**Reply by ivan-kleshnin** [reply]
>  I'm not trying to argue or anything, just wanting to understand.

Well, the author of the original solution explicitly said "Do not use it". It's an abuse of a type-system and overloads the machine even when it works. I'm 36 and I can recall many cases when something worked on my local machine and broke on QA where RAM/CPU can be limited (to reduce total expenses). It leads to hard-to-catch bugs and days wasted on debug sessions.

Looking from another side, I already wrap some zod validators for various reasons, one more is not a big deal.

**Reply by JacobWeisenburger** [reply]
I'm a self taught developer, so I'm just trying to learn because I haven't had the opportunity to work with a team yet. This isn't relevant to the original question, just wanting to give you some context about my questions.

Does QA normally run tests on typescript or the compiled javascript?

**Reply by ivan-kleshnin** [reply]
> I'm a self taught developer, so I'm just trying to learn because I haven't had the opportunity to work with a team yet

👌 hopefully we won't spam anyone here. After all, this Discussions are purposed for more casual-way dialogues.

> Does QA normally run tests on typescript or the compiled javascript?

1. By QA I mean a machine or a set of machines where QA team performs their acceptance tests (manual + automated).
2. As far as I know there's no way to run TypeScript directly. So it's always the compiled version that is run.
3. In our pipeline the build happens on QA machine. And it's a typical case IMO. We use NextJS framework. Each app build happens after both type-check and linting (NextJS defaults). So if type-check is slow or resource-heavy – it will cause problems. TS or ESLint errors will stop the pipeline and cancel the deployment process.

**Reply by JacobWeisenburger** [reply]
very interesting. thanks for the extra info. since I work by myself, I don't really know about things like this.

**Reply by JacobWeisenburger** [reply]
back to your original question. I don't know that there is another way to get the types the way you are asking without duplication. I am very interested to know if someone else has solved this problem or has any more insight.

**Reply by scotttrinh** [reply]
@ivan-kleshnin 

Could you go the other way and specify the `OPPORTUNITY_TYPES` first and then specify the `OPPORTUNITY_OPTIONS` as `Record<OpportunityTypes, string>` or something like that? Doesn't save any typing, but at least the compiler can be helpful if you are missing a case?

**Reply by ivan-kleshnin** [reply]
Yes, sounds like a good idea 💡
